File: java/src/main/java/cookbook/RecipeHttp.java
package cookbook;

import java.net.URL;
import java.net.HttpURLConnection;
import java.io.InputStreamReader;
import java.io.BufferedReader;
import java.util.Map;
import java.util.List;
import com.fasterxml.jackson.databind.ObjectMapper;
import com.fasterxml.jackson.core.type.TypeReference;
import java.net.MalformedURLException;
import java.net.ProtocolException;
import java.io.IOException;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;

public class RecipeHttp implements Recipe {

    private static final Logger LOGGER = LoggerFactory.getLogger(RecipeHttp.class);

    public void title() {

        LOGGER.debug("# Http");

    }

    public void cook() {

        String host = "weather.livedoor.com";
        String api = "/forecast/webservice/json/v1?city=130010";

        try {

            // HTTPリクエスト

            URL url = new URL("http://" + host + api);
            HttpURLConnection http = (HttpURLConnection)url.openConnection();
            http.setRequestMethod("GET");
            http.connect();

            BufferedReader reader = new BufferedReader(new InputStreamReader(http.getInputStream()));
            StringBuilder buf = new StringBuilder();
            String line;
            while ((line = reader.readLine()) != null) {
                buf.append(line);
            }

            // HTTPレスポンス解析

            ObjectMapper mapper = new ObjectMapper();
            Map<String, Object> data = mapper.readValue(buf.toString(), new TypeReference<Map<String, Object>>(){});

            @SuppressWarnings("unchecked")
            Map<String, Object> location = (Map<String, Object>)data.get("location");
            LOGGER.debug(String.format("%s, %s", location.get("prefecture"), location.get("city")));

            @SuppressWarnings("unchecked") 
            List<Map<String, Object>> forecasts = (List<Map<String, Object>>)data.get("forecasts");
            for(Map<String, Object> forecast: forecasts) {
                LOGGER.debug(String.format("%s[%s]: %s", forecast.get("dateLabel"), forecast.get("date"), forecast.get("telop")));
            }

        } catch (MalformedURLException ex) {
            LOGGER.error(ex.getMessage(), ex);
        } catch (ProtocolException ex) {
            LOGGER.error(ex.getMessage(), ex);
        } catch (IOException ex) {
            LOGGER.error(ex.getMessage(), ex);
        }

    }

    public static void main(String[] args) throws Exception {

        Recipe recipe = new RecipeHttp();

        recipe.title();
        recipe.cook();

    }

}

File: scala/src/main/scala/cookbook/RecipeHttp.scala
package cookbook

import scala.io.Source
import com.fasterxml.jackson.databind.ObjectMapper
import com.fasterxml.jackson.module.scala.DefaultScalaModule
import com.typesafe.scalalogging.Logger
import org.slf4j.LoggerFactory

class RecipeHttp private extends Recipe {

    val logger = Logger(LoggerFactory getLogger this.getClass)

    def title() = {

        logger debug "# Base"

    }

    def cook() = {

        val host = "weather.livedoor.com";
        val api = "/forecast/webservice/json/v1?city=130010";

        // HTTPリクエスト

        val src = Source fromURL "http://" + host + api
        var body: String = null
        try {
            body = src.getLines.mkString
        } finally {
            src.close
        }

        // HTTPレスポンス解析

        val mapper = new ObjectMapper
        mapper registerModule DefaultScalaModule
        val data = mapper.readValue(body, classOf[Map[String, _]])

        val location = data("location").asInstanceOf[Map[String, _]]
        logger debug s"${location("prefecture")}, ${location("city")}"

        val forecasts = data("forecasts").asInstanceOf[List[Map[String, _]]]
        for(forecast <- forecasts) {
            logger debug s"${forecast("dateLabel")}[${forecast("date")}]: ${forecast("telop")}"
        }

    }

}

object RecipeHttp {

    def apply() = {

        new RecipeHttp

    }

    def main(args: Array[String]) {

        val recipe = RecipeHttp.apply

        recipe.title
        recipe.cook

    }

}
